alacritty_terminal/src/tty/mod.rs

```diff
@@ -29,6 +30,10 @@ pub struct Options {
 
     /// Remain open after child process exits.
     pub hold: bool,
+
+    /// Extra environment variables.
+    #[cfg(not(windows))]
```

Comment:
Why we don't do so on windows?

Reply:
Mainly due to my 0 knowledge of how things work for Windows.

As far as I see, there are two options to spawn a terminal in Windows:
* poweshell commandline 
https://github.com/alacritty/alacritty/blob/8218c203032e6f42a20f31112c5e72abf6c3af83/alacritty_terminal/src/tty/windows/mod.rs#L125

would it be ok to use `$env1:value1;$env2:value2;..everything else` or what the approach would be

* `ConptyApi`
https://github.com/alacritty/alacritty/blob/8218c203032e6f42a20f31112c5e72abf6c3af83/alacritty_terminal/src/tty/windows/conpty.rs#L107

Presumably, `CreateProcessW` has to get more parameters?

Reply:
There's a command called `CreateProcessW` which we use, it has option to pass environment. Keep in mind that it sounds like `lpEnvironment` discards all the parent process envs, which we likely shouldn't be doing and rather merge/overwrite, but it needs testing.

It's the **only** way, the other one is the default thing which we shouldn't speculate with.

Reply:
I'm somewhat stuck with forming the right parameter form for the env variables.
Here's a simple example of an attempt to add some ad-hoc env variable as an argument

```diff
diff --git a/alacritty_terminal/src/tty/windows/conpty.rs b/alacritty_terminal/src/tty/windows/conpty.rs
index 9731b4f0..c3605316 100644
--- a/alacritty_terminal/src/tty/windows/conpty.rs
+++ b/alacritty_terminal/src/tty/windows/conpty.rs
@@ -1,9 +1,11 @@
 use log::info;
+use std::ffi::{OsStr, OsString};
 use std::io::Error;
+use std::os::windows::ffi::OsStrExt;
 use std::os::windows::io::IntoRawHandle;
 use std::{mem, ptr};
 
-use windows_sys::core::{HRESULT, PWSTR};
+use windows_sys::core::{IInspectable, HRESULT, PWSTR};
 use windows_sys::Win32::Foundation::{HANDLE, S_OK};
 use windows_sys::Win32::System::Console::{
     ClosePseudoConsole, CreatePseudoConsole, ResizePseudoConsole, COORD, HPCON,
@@ -201,6 +203,24 @@ pub fn new(config: &Options, window_size: WindowSize) -> Option<Pty> {
     let cmdline = win32_string(&cmdline(config));
     let cwd = config.working_directory.as_ref().map(win32_string);
 
+    let env_block_wide: Vec<u16> = //config.env
+        Some(("FOO", "FOO_VALUE"))
+        .iter()
+        .flat_map(|(key, value)| {
+            let wide_string: Vec<u16> = OsStr::new(key)
+                .encode_wide()
+                .chain(std::iter::once(b'=' as u16))
+                .chain(OsStr::new(value).encode_wide())
+                .chain(std::iter::once(0)) // Null terminator for the individual string
+                .collect();
+            wide_string
+        })
+        .chain(std::iter::once(0)) // Add an extra null terminator to the end
+        .collect();
+
+    let another_env_block_wide =
+        another::format_env_block(&[(OsString::from("FOO"), OsString::from("FOO_VALUE"))]);
+
     let mut proc_info: PROCESS_INFORMATION = unsafe { mem::zeroed() };
     unsafe {
         success = CreateProcessW(
@@ -210,7 +230,9 @@ pub fn new(config: &Options, window_size: WindowSize) -> Option<Pty> {
             ptr::null_mut(),
             false as i32,
             EXTENDED_STARTUPINFO_PRESENT,
-            ptr::null_mut(),
+            Some(another_env_block_wide).as_ref().map(|a| a.as_ptr()).unwrap_or(ptr::null())
+                as IInspectable,
+            // ptr::null_mut(),
             cwd.as_ref().map_or_else(ptr::null, |s| s.as_ptr()),
             &mut startup_info_ex.StartupInfo as *mut STARTUPINFOW,
             &mut proc_info as *mut PROCESS_INFORMATION,
@@ -249,3 +271,43 @@ impl From<WindowSize> for COORD {
         COORD { X: columns as i16, Y: lines as i16 }
     }
 }
+
+mod another {
+    use std::{
+        collections::HashSet,
+        ffi::{OsStr, OsString},
+        os::windows::ffi::{OsStrExt, OsStringExt},
+    };
+
+    pub fn format_env_block(env: &[(OsString, OsString)]) -> Vec<u16> {
+        fn to_uppercase(s: &OsStr) -> OsString {
+            OsString::from_wide(
+                &s.encode_wide()
+                    .map(
+                        |c| {
+                            if c < 128 {
+                                (c as u8 as char).to_ascii_uppercase() as u16
+                            } else {
+                                c
+                            }
+                        },
+                    )
+                    .collect::<Vec<_>>(),
+            )
+        }
+        let mut pruned: Vec<_> = {
+            let mut seen = HashSet::<OsString>::new();
+            env.iter().rev().filter(|&(k, _)| seen.insert(to_uppercase(k))).collect()
+        };
+        pruned.reverse();
+        let mut block = vec![];
+        for (k, v) in pruned {
+            block.extend(k.encode_wide());
+            block.push('=' as u16);
+            block.extend(v.encode_wide());
+            block.push(0);
+        }
+        block.push(0);
+        block
+    }
+}
```


with the `mod another` using https://github.com/hniksic/rust-subprocess/blob/5e89ac093f378bcfc03c69bdb1b4bcacf4313ce4/src/popen.rs#L1080

`Some(another_env_block_wide).as_ref().map(|a| a.as_ptr()).unwrap_or(ptr::null())` contraption also comes from that code, but I did try a regular version too with the same result:

```
panicked at alacritty_terminal\src\tty\windows\conpty.rs:257:5:
Unable to spawn shell: The parameter is incorrect. (os error 87)
error: process didn't exit successfully: target\debug\alacritty.exe (exit code: 0xc000013a, STATUS_CONTROL_C_EXIT)
```

So looks like I convert the parameter incorrectly, but I do not understand why does not it work for `rust-subprocess` crate code too.
Will poke around more, but really appreciate any pointers.

Reply:
And if I add the `|= CREATE_UNICODE_ENVIRONMENT` to the `dwcreationflags`, it does not output the error, but very briefly opens and closes the process, which also seems wrong..

Reply:
Added, thank you for the pointers.

https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa#parameters

* we do indeed inherit the env on default
> If this parameter is NULL, the new process uses the environment of the calling process.

* for non-defaults, have to deduplicate the variables, case-insensitive
https://learn.microsoft.com/sk-SK/previous-versions/troubleshoot/windows/win32/createprocess-cannot-eliminate-duplicate-variables#environment-variables

* I've cast the pointer to `IInspectable` logical type, but not sure if it's applicable, but other `*mut c_void` types there neither look applicable.

Seems to work:
![telegram-cloud-photo-size-2-5407061907099212075-y](https://github.com/alacritty/alacritty/assets/2690773/43a4a7c3-4cd5-4784-92f6-2a8b9ce3c3a7)